pkg/runtime/interfaces.go

```diff
@@ -23,64 +23,74 @@ import (
-	// TODO: Remove this method?
-	DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, groupVersionKind unversioned.GroupVersionKind) error
-
-	DecodeParametersInto(parameters url.Values, obj Object) error
+// Typer retrieves information about an object's group, version, and kind.
+type Typer interface {
+	// ObjectKind returns the version and kind of the provided object, or an
+	// error if the object is not recognized (IsNotRegisteredError will return true).
+	// It returns whether the object is considered unversioned at the same time.
+	ObjectKind(Object) (*unversioned.GroupVersionKind, bool, error)
 }
 
-// Encoder defines methods for serializing API objects into bytes
 type Encoder interface {
-	Encode(obj Object) (data []byte, err error)
-	EncodeToStream(obj Object, stream io.Writer) error
+	// EncodeToStream writes an object to a stream. Override versions may be provided for each group
+	// that enforce a certain versioning. Implementations may return errors if the versions are incompatible,
+	// or if no conversion is defined.
+	EncodeToStream(obj Object, stream io.Writer, overrides ...unversioned.GroupVersion) error
+}
 
-	// TODO: Add method for processing url parameters.
-	// EncodeParameters(obj Object) (url.Values, error)
+type Decoder interface {
+	// Decode attempts to deserialize the provided data using either the innate typing of the scheme or the
+	// default kind, group, and version provided. It returns a decoded object as well as the actual kind, group, and
+	// version decoded, or an error. If into is non-nil, it will be used as the target type and implementations
+	// may choose to use it rather than reallocating an object. However, the object is not guaranteed to be
+	// populated. The returned object type is not guaranteed to match into.
+	Decode(data []byte, defaults *unversioned.GroupVersionKind, into Object) (Object, *unversioned.GroupVersionKind, error)
 }
 
-// ObjectCodec represents the common mechanisms for converting to and from a particular
-// binary representation of an object.
-// TODO: Remove this interface - it is used only in CodecFor() method.
-type ObjectCodec interface {
+// Serializer is the core interface for transforming objects into a serialized format and back.
+// Implementations may choose to perform conversion of the object, but no assumptions should be made.
+type Serializer interface {
+	Encoder
 	Decoder
+}
 
-	// EncodeToVersion convert and serializes an object in the internal format
-	// to a specified output version. An error is returned if the object
-	// cannot be converted for any reason.
-	EncodeToVersion(obj Object, outVersion string) ([]byte, error)
-	EncodeToVersionStream(obj Object, outVersion string, stream io.Writer) error
+// Codec is a Serializer that deals with the details of versioning objects. It offers the same
+// interface as Serializer, so this is a marker to consumers that care about the version of the objects
+// they receive.
+type Codec Serializer
```

Comment:
Clarify usage.  Serializer is passed to interfaces that expect raw
behavior.  Codec is passed to things that expect conversion.  There are
other factorings in the future, such as having Decode take more options
(like the selected serializer).  I don't want to hang up on that - the
changes I am making to isolate clients from the impact of this makes
Serializer a lower level stratification, not a higher level one.

On Thu, Dec 10, 2015 at 8:56 AM, David Eads <notifications@github.com>
wrote:

> In pkg/runtime/interfaces.go
> <https://github.com/kubernetes/kubernetes/pull/17922#discussion_r47227948>
> :
>
> >
> > -	// EncodeToVersion convert and serializes an object in the internal format
> > -	// to a specified output version. An error is returned if the object
> > -	// cannot be converted for any reason.
> > -	EncodeToVersion(obj Object, outVersion string) ([]byte, error)
> > -	EncodeToVersionStream(obj Object, outVersion string, stream io.Writer) error
> > +// Codec is a Serializer that deals with the details of versioning objects. It offers the same
> > +// interface as Serializer, so this is a marker to consumers that care about the version of the objects
> > +// they receive.
> > +type Codec Serializer
>
> Even with the comment, the utility of this interface isn't clear to me.
> Why shouldn't I just check for Serializer? Do we anticipate drift?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17922/files#r47227948>.
>